test/integration/sds_dynamic_integration_test.cc

```diff
@@ -40,17 +46,56 @@ namespace Ssl {
 // Hack to force linking of the service: https://github.com/google/protobuf/issues/4221.
 const envoy::service::secret::v3::SdsDummy _sds_dummy;
 
+std::string sdsTestParamsToString(
+    const ::testing::TestParamInfo<std::tuple<Network::Address::IpVersion, Grpc::ClientType, bool>>&
+        p) {
+  return fmt::format(
+      "{}_{}_{}", std::get<0>(p.param) == Network::Address::IpVersion::v4 ? "IPv4" : "IPv6",
+      std::get<1>(p.param) == Grpc::ClientType::GoogleGrpc ? "GoogleGrpc" : "EnvoyGrpc",
+      std::get<2>(p.param) ? "UsesQuic" : "UsesTcp");
+}
+
+std::vector<std::tuple<Network::Address::IpVersion, Grpc::ClientType, bool>>
+getSdsTestsParams(bool test_quic) {
+  std::vector<std::tuple<Network::Address::IpVersion, Grpc::ClientType, bool>> ret;
+  for (auto ip_version : TestEnvironment::getIpVersionsForTest()) {
+    for (auto grpc_type : TestEnvironment::getsGrpcVersionsForTest()) {
+      ret.push_back({ip_version, grpc_type, false});
+      if (test_quic) {
+#ifdef ENVOY_ENABLE_QUIC
+        ret.push_back({ip_version, grpc_type, true});
+#else
+        ENVOY_LOG_MISC(warn, "Skipping HTTP/3 as support is compiled out");
+#endif
+      }
+    }
+  }
+  return ret;
+}
+
 // Sds integration base class with following support:
 // * functions to create sds upstream, and send sds response
 // * functions to create secret protobuf.
-class SdsDynamicIntegrationBaseTest : public Grpc::GrpcClientIntegrationParamTest,
-                                      public HttpIntegrationTest {
+class SdsDynamicIntegrationBaseTest
+    : public Grpc::BaseGrpcClientIntegrationParamTest,
+      public HttpIntegrationTest,
+      public testing::TestWithParam<
+          std::tuple<Network::Address::IpVersion, Grpc::ClientType, bool>> {
```

Comment:
Maybe document what the gRPC test parameter is used for? Also what the bool refer to

Reply:
I change the tuple to a struct. Hope it helps readability.